### PR TITLE
BCDA-2005 Maintenance: DB Schema Migration Strategy

### PR DESCRIPTION
**Pull request by dhgreene**
### Fixes [BCDA-2005](https://jira.cms.gov/browse/BCDA-2005)
Up to this point, SSAS has been using GORM automigration to maintain the database schema.  This is a somewhat opaque, one-way process that works well in development environments, but is less desirable in production.

This PR introduces the database schema migration tool `golang/migrate`, which uses SQL scripts to make changes and to roll them back.

### Proposed Changes
- Call `golang/migrate` instead of GORM automigrate for all database schema changes
- Begin writing SQL scripts to make changes and roll them back
- Begin testing SQL migrations as part of `make test`

### Change Details
- A database dump of the current SSAS tables/relations/indices constitutes the first migration
- `make migrations-test` runs migration tests, and is run at the end of `make test`

### Security Implications
- [x] new software dependencies
A new executable (`migrate`) is added to local containers that need to make database schema changes.  How to use this tool in AWS environments will be covered by a separate PR.

- [ ] security controls or supporting software altered
- [ ] new data stored or transmitted
- [x] security checklist is completed for this change
[Security Checklist - SSAS DB Schema Change Strategy](https://confluence.cms.gov/display/BCDA/Security+Checklist+-+SSAS+DB+Schema+Change+Strategy)
- [ ] requires more information or team discussion to evaluate security implications

### Acceptance Validation
Screen Shot 2019-10-15 at 10 53 21 AM

### Feedback Requested
- Actively soliciting optimizations to the Docker images and invocations
- Is the testing strategy sound?
- Does the README.md need more content?  What kind?
- All other comments welcome

### CODE REVIEW

File: ops/migrations_test.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#
+# This script is intended to be run from within the Docker "unit_test" container
+# The docker-compose file brings forward the env vars: DB
+#
+set -e
+set -o pipefail
+
+timestamp=`date +%Y-%m-%d_%H-%M-%S`
+
+echo "Setting up test DB (bcda_test)..."
```
Comment (msnook) [inline ops/migrations_test.sh]: Should the test db be named `ssas_test` in this env?

File: unit_test.sh [hunk 2]
```diff
@@ -18,9 +18,9 @@ echo "TEST_DB_URL is $TEST_DB_URL"
 usql $DB_HOST_URL -c 'drop database if exists bcda_test;'
 usql $DB_HOST_URL -c 'create database bcda_test;'
```
Comment (msnook) [inline unit_test.sh]: `ssas_test` instead of `bcda_test`?

File: Dockerfiles/Dockerfile.tests [hunk 3]
```diff
@@ -13,6 +13,12 @@ COPY . .
 RUN dep ensure
 
 RUN apt-get update
+RUN apt-get install apt-transport-https ca-certificates libunwind8 icu-devtools python-pip -y
```
Comment (msnook) [inline Dockerfiles/Dockerfile.tests]: Does `golang-migrate` need to exist in the test image?  Could the `migrations_test.sh` simply be triggered via CLI from the ssas container?

File: Makefile [hunk 4]
```diff
@@ -29,6 +29,9 @@ postman:
 	# For example: make postman env=test token=<MY_TOKEN>
 	docker-compose -f docker-compose.test.yml run --rm postman_test test/postman_test/SSAS.postman_collection.json -e test/postman_test/ssas-local.postman_environment.json --global-var adminClientId=$(SSAS_ADMIN_CLIENT_ID) --global-var adminClientSecret=$(SSAS_ADMIN_CLIENT_SECRET)
 
+migrations-test:
+	docker-compose -f docker-compose.test.yml run --rm tests bash ops/migrations_test.sh
```
Comment (msnook) [inline Makefile]: Could this just be run via CLI from the SSAS container?  Is it pointed at the same DB in both cases?  If so, I would argue that `golang-migrate` can be removed from the tests image.

File: db/migrations/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Database Schema Changes
+
+## Philosophy
+* All database changes will be modeled in discrete schema migrations
+    * Code to roll back a migration will be included unless highly impracticable
+* Migrations will receive unique sequential integer identifiers.
+    * Engineers should keep aware of identifier collisions (and in some instances required migration order), and adjust their migration ID’s before merging with master, if necessary.
```
Comment (msnook) [inline db/migrations/README.md]: If a migration is run on `0001` and a few days later `0002` is written, what happens if I run `up`?  Will both `0001` and `0002` be executed?  Or is it smart enough to know that only `0002` needs to be run?

**Comment by msnook** [comment]
Good stuff.  I'm curious how we will transition to using this in AWS.  We cannot start at `0001` in AWS, since we already have our DB in place.

File: ops/migrations_test.sh (hunk 1 above)
Comment (dhgreene) [reply ops/migrations_test.sh]: Probably.  But the main database is still `bcda`, so I'm not sure whether `bcda_test` is more congruent.

File: Dockerfiles/Dockerfile.tests (hunk 3 above)
Comment (dhgreene) [reply Dockerfiles/Dockerfile.tests]: It needs to be there for the migrations test, which does not run in the ssas container.

File: Makefile (hunk 4 above)
Comment (dhgreene) [reply Makefile]: No, the SSAS container points at the non-test DB.

File: db/migrations/README.md (hunk 5 above)
Comment (dhgreene) [reply db/migrations/README.md]: It's smart enough to know, by keeping track in a table named `database_migrations`.

File: ops/migrations_test.sh (hunk 1 above)
Comment (msnook) [reply ops/migrations_test.sh]: Yeah, makes sense to leave it as is.

**Comment by dhgreene** [comment]
@msnook I've made changes based on our conversation on Friday, and this is ready for re-review.

**Pull request merged**